### PR TITLE
Generating auditd rules is now optional, handled by auditd_apply_audit_rules variable

### PR DESCRIPTION
**Pull request by polachz**
As mentioned in issue #42 aplying auditd hardening rules is now driven by the `auditd_apply_audit_rules` variable. Documentation is changed as well.

During this modification, I have discovered that kernel parameters are modified by direct way, what can be problem if another task will try to add or remove any other kernel parameter - the $GRUB_CMDLINE_LINUX variable will not reflect new state, I think. I have modified this to RHEL/CentOS recommended way, use the **grubby** tool. It works smoothly and modify grubenv file by these parameters...

Next fixed issue is direct modification of the _/usr/lib/systemd/system/auditd.service_ file. The _"black magic"_ used here is no more necessary onto the CentOS/RHEL v 8 and v7. Here is enough to place rules to /etc/audit/rules.d directory. then auditd provide all other tasks itself. Verified by **auditctl -l** command -> rules are correctly applied.

### CODE REVIEW

**Comment by konstruktoid** [comment]
Looks good, I'll test it a bit, but could you fix the two linting issues that's present?

**Comment by polachz** [comment]
> Looks good, I'll test it a bit, but could you fix the two linting issues that's present?

Seems that Lint is happy now :)

File: tasks/auditd.yml [hunk 1]
```diff
@@ -18,41 +18,14 @@
...
-  lineinfile:
-    line: 'GRUB_CMDLINE_LINUX="$GRUB_CMDLINE_LINUX {{ grub_audit_cmdline }} {{ grub_audit_backlog_cmdline }}"'
-    dest: /etc/default/grub
-    insertafter: '^GRUB_CMDLINE_LINUX="[a-z]+'
-    state: present
-    create: 'no'
-    mode: 0644
-    owner: root
-    group: root
+  command: grubby --update-kernel=ALL --args="{{ grub_audit_cmdline }} {{ grub_audit_backlog_cmdline }}"
```
Comment (konstruktoid) [inline tasks/auditd.yml:21]: This is not a idempotent task, and needs a check so it isn't executed if `{{ grub_audit_cmdline }} {{ grub_audit_backlog_cmdline }}` is present in `GRUB_CMDLINE_LINUX`.

File: tasks/auditd.yml [hunk 2]
```diff
@@ -18,41 +18,14 @@
...
-- name: check RedHat auditd ExecStartPost
-  environment:
-    PATH: /usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
-  command: grep -qo "^ExecStartPost=-/sbin/auditctl -R" /usr/lib/systemd/system/auditd.service
-  register: execstartpost_auditd_set
-  when: ansible_os_family == "RedHat"
-  changed_when: execstartpost_auditd_set.rc != 0
-  failed_when: execstartpost_auditd_set.rc > 1
-
-- name: set audit service execstartpost
```
Comment (konstruktoid) [inline tasks/auditd.yml:46]: why is this removed?

File: tasks/auditd.yml (hunk 1 above)
Comment (polachz) [reply tasks/auditd.yml:21]: Grubby grants options presence. If already present, then will be not added. If missing, grub will add. 

Additionally, in original task, the mentioned check is missing, isn't it ?  So it can be duplicated by old way. The new one with grubby grants no duplicity

File: tasks/auditd.yml (hunk 2 above)
Comment (polachz) [reply tasks/auditd.yml:46]: Used only in next removed task:  name: set audit service execstartpost

I have provided explanation in PR first comment,...

Comment (konstruktoid) [reply tasks/auditd.yml:46]: why not? https://github.com/konstruktoid/ansible-role-hardening/blob/master/tasks/auditd.yml#L37-L55 ensures that the current rules are loaded when `auditd` is (re)started.

File: tasks/auditd.yml (hunk 1 above)
Comment (konstruktoid) [reply tasks/auditd.yml:21]: https://github.com/konstruktoid/ansible-role-hardening/blob/master/tasks/auditd.yml#L22 is idempotent since `lineinfile` wont change any file if the line is present.

that might be so, but `ansible` doesn't agree.

```shell
$ ansible-playbook -v -i '127.0.0.1,' -c local audit_grub.yml 
Using /etc/ansible/ansible.cfg as config file

PLAY [all] *************************************************************************************************************************************************************************************************

TASK [Gathering Facts] *************************************************************************************************************************************************************************************
ok: [127.0.0.1]

TASK [set RedHat audit grub cmdline] ***********************************************************************************************************************************************************************
changed: [127.0.0.1] => {"changed": true, "cmd": ["grubby", "--update-kernel=ALL", "--args=audit_backlog_limit=8192 audit=1"], "delta": "0:00:00.177467", "end": "2021-02-16 13:01:23.498124", "rc": 0, "start": "2021-02-16 13:01:23.320657", "stderr": "", "stderr_lines": [], "stdout": "", "stdout_lines": []}

PLAY RECAP *************************************************************************************************************************************************************************************************
127.0.0.1                  : ok=2    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   

$ ansible-playbook -v -i '127.0.0.1,' -c local audit_grub.yml 
Using /etc/ansible/ansible.cfg as config file

PLAY [all] *************************************************************************************************************************************************************************************************

TASK [Gathering Facts] *************************************************************************************************************************************************************************************
ok: [127.0.0.1]

TASK [set RedHat audit grub cmdline] ***********************************************************************************************************************************************************************
changed: [127.0.0.1] => {"changed": true, "cmd": ["grubby", "--update-kernel=ALL", "--args=audit_backlog_limit=8192 audit=1"], "delta": "0:00:00.176944", "end": "2021-02-16 13:01:28.831339", "rc": 0, "start": "2021-02-16 13:01:28.654395", "stderr": "", "stderr_lines": [], "stdout": "", "stdout_lines": []}

PLAY RECAP *************************************************************************************************************************************************************************************************
127.0.0.1                  : ok=2    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0 
```

Comment (polachz) [reply tasks/auditd.yml:21]: Oh I see. The grubby doesn't report if changed something or not.  But it add parameters only once. if already exists, it's not duplicit them. I have checked this again. 

The grubby is recommended way how to modify grub2 entries on RHEL now.... 

Provide an detection before the task can be really complicated (due possible several kernel configurations).

Is this behavior real problem ?

Comment (konstruktoid) [reply tasks/auditd.yml:21]: Define a "real problem".

https://cumulusnetworks.com/blog/netdevops-important-idempotence/
https://molecule.readthedocs.io/en/latest/faq.html#why-is-my-idempotence-action-failing

File: tasks/auditd.yml (hunk 2 above)
Comment (polachz) [reply tasks/auditd.yml:46]: 1. The rule modifies the **/usr/lib/systemd/system/auditd.service** Then any updat will kill the change 
2. in the rules is final rule **-e 2**  forces audit to be put in immutable mode. Audit changes can only be made on system reboot.
3. Seems that staert the audit service again forces rules re-generation... 

By my opinion, it's obsolete, and if not, then it's implemented on the wrong place

Comment (konstruktoid) [reply tasks/auditd.yml:46]: Yeah, `/usr/lib/systemd/system/auditd.service` is the incorrect file. It should be placed in `/etc`.

Update (the wrong) .service file at https://github.com/konstruktoid/ansible-role-hardening/blob/master/tasks/auditd.yml#L50, add our rules at https://github.com/konstruktoid/ansible-role-hardening/blob/master/tasks/auditd.yml#L165 and then generate and restart `auditd` if file has been changed at https://github.com/konstruktoid/ansible-role-hardening/blob/master/handlers/main.yml#L2-L12.  
This is to ensure our rules are in place and used.

Comment (konstruktoid) [reply tasks/auditd.yml:46]: >     3. Seems that staert the audit service again forces rules re-generation...
where?

File: tasks/auditd.yml (hunk 1 above)
Comment (polachz) [reply tasks/auditd.yml:21]: Thhat if the task report the change, even if all paremeters are really set. is it a problem? 

because if you use this:
line: 'GRUB_CMDLINE_LINUX="$GRUB_CMDLINE_LINUX {{ grub_audit_cmdline }} {{ grub_audit_backlog_cmdline }}"'

then by my opinion

1. If only one parameter is already set, then no change is provided
2. If another taks will provide something simmilar 'GRUB_CMDLINE_LINUX="$GRUB_CMDLINE_LINUX param1 param2" then previous change will be lost, because who upates the $GRUB_CMDLINE_LINUX variable? 

I think that lost changes or do not get them all is more serious then this one with change detection...

Comment (konstruktoid) [reply tasks/auditd.yml:21]: `grubby`  is the correct way to update it on CentOS/RHEL, since we can change this on Ubuntu et al in the `grub.d` directory.
what is needed is that if `{{ grub_audit_cmdline }} {{ grub_audit_backlog_cmdline }}` is already present, `grubby` shouldn't be used. this is to ensure nothing is changed unless it has to be.
I know the file itself doesn't change, but as long as ansible considers otherwise, the role isn't idempotent.

this is a small issue, and I can fix it after merge.

File: tasks/auditd.yml (hunk 2 above)
Comment (polachz) [reply tasks/auditd.yml:46]: Sorry. Point3 is wrong. but with -e 2 as last rule, you can reload, restart, re-generat rules, but without any efect. You have to reboot the machine to modify the rule

File: tasks/auditd.yml (hunk 1 above)
Comment (polachz) [reply tasks/auditd.yml:21]: The problem with detection here is, that sometimes it can be in grubenv file, sometimes somewhere else, depends if EFI or BIOS is used, etc :(

maybe the  `grub2-editenv - list | grep kernelopts` can be used and then pres the output. But it's really lot of work for small detection.. :(

File: tasks/auditd.yml (hunk 2 above)
Comment (konstruktoid) [reply tasks/auditd.yml:46]: yeah, should probably use a `when:` if `-e 2` is set

Comment (konstruktoid) [reply tasks/auditd.yml:46]: will fix those tasks after merge so you dont have to mess around with rebasing

**Comment by konstruktoid** [comment]
Will merge and fix the issues
